src/server/ApiDispatchers.cpp

```diff
@@ -1252,38 +1252,35 @@ static DWORD TraceGetThreadId(CONSOLE_API_MSG* const m)
     ULONG cbBufferSize;
     RETURN_IF_FAILED(m->GetInputBuffer(&pvBuffer, &cbBufferSize));
 
-    PVOID pvInputTarget;
-    const auto cbInputTarget = a->TargetLength;
-    PVOID pvInputExeName;
-    const auto cbInputExeName = a->ExeLength;
-    PVOID pvInputSource;
-    const auto cbInputSource = a->SourceLength;
-    // clang-format off
-    RETURN_HR_IF(E_INVALIDARG, !IsValidStringBuffer(a->Unicode,
-                                                    pvBuffer,
-                                                    cbBufferSize,
-                                                    3,
-                                                    cbInputExeName,
-                                                    &pvInputExeName,
-                                                    cbInputSource,
-                                                    &pvInputSource,
-                                                    cbInputTarget,
-                                                    &pvInputTarget));
-    // clang-format on
+    // There are 3 strings stored back-to-back within the message payload.
+    // First we verify that their size and alignment are alright and then we extract them.
+    const ULONG cbInputExeName = a->ExeLength;
+    const ULONG cbInputSource = a->SourceLength;
+    const ULONG cbInputTarget = a->TargetLength;
+
+    const auto alignment = a->Unicode ? alignof(wchar_t) : alignof(char);
+    // ExeLength, SourceLength and TargetLength are USHORT and summing them up will not overflow a ULONG.
+    const auto badLength = cbInputTarget + cbInputExeName + cbInputSource > cbBufferSize;
+    const auto badAlignment = ((cbInputExeName | cbInputSource | cbInputTarget) & (alignment - 1)) != 0;

```

Comment:
what the shit

Reply:
"the buffer has a bad alignment if the LSB of (the input buffers) & (the inverse of the LSB of alignment) is not zero"

There's like a lot of magic there. 
* The first expression in parens checks the char vs wchar size. It makes sure the buffers are all the either even sized. If one is odd, they all count as char buffers. 
* The second expression does bit magic to convert `alignof(wchar_t)==0b0010` to `0b0001`, and `alignof(char)==0b0001` -> `0b0000`. 
* Then do a bitwise AND. That'll be 0 only if
  - all the buffers are even sized and we asked for a wchar OR char buffer
  - at least one is odd and we asked for a char buffer
  
  
okay I followed it through, but damn this is bit magic

Reply:
Well... There are 3 options I can think of: https://godbolt.org/z/fjWrEhdvo
For me personally they're all 3 about the same. I often use the `& (x - 1)` pattern for things like aligning pointers to powers of 2. Hence me picking the first option.